test/omega_red/cache_test.clj

```diff
@@ -0,0 +1,63 @@
+(ns omega-red.cache-test
+  (:require
+    [clojure.test :refer [deftest testing is use-fixtures]]
+    [com.stuartsierra.component  :as component]
+    [omega-red.cache :as cache]
+    [omega-red.protocol :as redis]
+    [omega-red.redis]))
+
+
+(def redis-config
+  {:host (or (System/getenv "REDIS_URL") "127.0.0.1")
+   :port (Integer/parseInt (or (System/getenv "REDIS_PORT") "6379"))})
+
+
+(def system
+  {:redis (omega-red.redis/create redis-config)
+   :cache (component/using
+            (cache/create)
+            [:redis])})
+
+
+(def state (atom 0))
+
+(def sys (atom nil))
+
+
+(defn stateful []
+  (swap! state inc))
+
+
+(use-fixtures :once (fn [test]
+                      (reset! state 0)
+                      (reset! sys (component/start (component/map->SystemMap system)))
+                      (redis/execute (:redis @sys) [:del "testing:1"])
+                      (test)
+                      (swap! sys component/stop)))
+
+
+(deftest cache-test
+  (testing "caches result of fetch call"
+    (is (= 1 (cache/fetch-and-set (:cache @sys) {:fetch stateful
+                                                 :cache-get [:get "testing:1"]
+                                                 :cache-set (fn [cache-result]
+                                                              [:set "testing:1" cache-result])})))
+    (is (= 2 (stateful)))
+    (is (= 3 (stateful)))
+    ;; hmmmmm, this is not helpful
+    (is (= "1" (cache/fetch-and-set (:cache @sys) {:fetch stateful
```

Comment:
The naming looks ok being used this way, I didn't realize `cache-get/set` will be vector of args for redis fn.

Reply:
Yeah, it feels a bit magical... Maybe something like this would work in more obvious way, but it would require moving protocols and components around to allow for nicer API:

```clojure
(redis/cache-fetch-and-set redis {:fetch some-fetch-fn
                                  :cache-set (fn [redis fetch-result]) ;; you have the redis connection in callback 
                                  :cache-get (fn [redis] );; you also get the connection in cache read callback
 })
```

a bit less magical and more explicit.

cc @AfonsoTsukamoto @mjou-uz

Reply:
I like this last API interface, just two considerations:

The redis connection... cannot be set previously as a partial? Then the caller does not have to import redis in the namespace


I will also create a simpler `get-and-set` to use `get` and `setx` redis command, that will cover majority of cases and will lead to cleaner calls.
That will be then:
```
(redis/getand-set redis {:fetch some-fetch-fn 
                         :key "some-cool-key"
                         :exp XX  ;; expiration time})
```

Reply:
1. You have to pass the conneciton/component to something to call a protocol function on it, so that's unavoidable. But yes, I don't like separate namespaces - let me see if merging protocols and adding caching behavior to the Redis component looks better
2. That's the thing - we shouldn't limit ourselves to the basic `setx` and `get` commands, internally we use different data structures for caches, for example sets and hash-maps - and these require different commands for reading and updating ( `hset` and `hget` / `hgetall` - in case of maps). So we can keep this a bit more open and just tidy up the overall logic.


Let me push an update to address the comments so far 👍